### PR TITLE
Prefs [Task][26305] Create Check In screen UI

### PR DESCRIPTION
**Pull request by toandv98**
## Related Tickets

- [#Task 26305](https://edu-redmine.sun-asterisk.vn/issues/26305)

## WHAT

- Create Check In screen UI.

## HOW

- Switch button check in, TextView time in, Button back, AnimatedVectorDrawable

## Review Checklist

| Category      | View Point                                                   | Description                                                  | Expected Reviewer Answer | Reviewer1 (ToanDV) | Reviewer2 (HuyTQ)   |
| ------------- | ------------------------------------------------------------ | ------------------------------------------------------------ | ------------------------ | ------------------ | ------------------ |
| General       | Code should be self-explanatory.                             | Get a feel of story reading, while going through the code. Use appropriate name for variables, functions and classes. If you are taking more time to understand the code, then either code needs refactoring or at least comments have to be written to make it clear. | YES                      | <li>- [x] yes</li> | <li>- [x] yes</li> |
| General       | Is this function or class too big? If yes, is the function or class having too many responsibilities? |                                                              | NO                       | <li>- [x] no</li>  | <li>- [x] no</li>  |
| General       | Does the code follow DRY?: Do not Repeat Yourself            | Is there any redundant or duplicate code? DRY (Do not Repeat Yourself) principle: The same code should not be repeated more than twice. Consider reusable services, functions and components. Consider generic functions and classes. | YES                      | <li>- [x] yes</li> | <li>- [x] yes</li> |
| General       | Exception handling and cleanup (dispose) resources           |                                                              | YES                      | <li>- [x] yes</li> | <li>- [x] yes</li> |
| General       | Does the code build and execute correctly?                   |                                                              | YES                      | <li>- [x] yes</li> | <li>- [x] yes</li> |
| Conventions   | Does the source code have end-line errors?                   |                                                              | NO                       | <li>- [x] no</li>  | <li>- [x] no</li>  |
| Conventions   | No hard coding, use constants/configuration values           |                                                              | YES                      | <li>- [x] yes</li> | <li>- [x] yes</li> |
| Conventions   | Does the code follow Sun* coding style and coding conventions? | https://github.com/framgia/coding-standards/tree/master/eng/android | YES                      | <li>- [x] yes</li> | <li>- [x] yes</li> |
| Redmine       | Does the ticket follow Sun* Redmine working process?         | https://github.com/framgia/Training-Guideline/blob/master/WorkingProcess/redmine/redmine.md | YES                      | <li>- [x] yes</li> | <li>- [x] yes</li> |
| Documentation | Is there any incomplete code? If so, should it be removed or flagged with a suitable marker like ‘TODO’? |                                                              | YES                      | <li>- [x] yes</li> | <li>- [x] yes</li> |

## Notes (Optional)

*(Impacted Areas in Application(List features, api, models or services that this PR will affect))*

*(List gem, library third party add new)*

*(Other notes)*

### CODE REVIEW

File: app/src/main/res/layout/dialog_checkin.xml [hunk 1]
```diff
@@ -0,0 +1,84 @@
+<?xml version="1.0" encoding="utf-8"?>
+<androidx.constraintlayout.widget.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    android:id="@+id/linearLayout"
+    android:layout_width="match_parent"
+    android:layout_height="match_parent"
+    android:background="@color/color_bg_test"
```
Comment (tuanpa-0908) [inline app/src/main/res/layout/dialog_checkin.xml]: color test là màu gì thế?

File: app/src/main/java/com/edu/sun/oereminder/ui/timetracking/TimeRecordAdapter.kt [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    }
+
+    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
+        holder.bind(getItem(position))
+    }
+
+    class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
+        fun bind(timeRecord: TimeRecord) {
+            with(itemView) {
+                with(timeRecord) {
```
Comment (thonc-2135) [inline app/src/main/java/com/edu/sun/oereminder/ui/timetracking/TimeRecordAdapter.kt:30]: có thể with(itemView.timeRecord) { .....được không nhỉ

File: app/src/main/java/com/edu/sun/oereminder/ui/timetracking/TimeTrackingPresenter.kt [hunk 3]
```diff
@@ -8,4 +8,12 @@ import com.edu.sun.oereminder.ui.timetracking.TimeTrackingContract.View
 class TimeTrackingPresenter(private val timeSheetRepository: TimeSheetRepository) :
     BasePresenterImpl<View>(), Presenter {
 
+    override fun load() {
+
```
Comment (thonc-2135) [inline app/src/main/java/com/edu/sun/oereminder/ui/timetracking/TimeTrackingPresenter.kt]: xóa dòng trống đi, máy cái này viết 1 dòng cũng được, việc gì phải thêm  nhiều dòng thế.
`The first rule of functions is that they should be small. The second rule of functions is that they should be smaller than that
`

File: app/src/main/java/com/edu/sun/oereminder/utils/DateTimeExtension.kt [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+import java.util.*
+
+const val DATE_FORMAT = "dd MMMM"
+const val DATE_FORMAT_DAY_OF_WEEK = "EEEE, dd-MM-yyyy"
+const val TIME_FORMAT = "hh:mm"
+const val TIME_NULL = "--:--"
+
+fun GregorianCalendar.toDate(): String =
+    SimpleDateFormat(DATE_FORMAT_DAY_OF_WEEK, Locale.getDefault()).format(time)
+
```
Comment (thonc-2135) [inline app/src/main/java/com/edu/sun/oereminder/utils/DateTimeExtension.kt:13]:
```suggestion
fun GregorianCalendar.toDate() =
    SimpleDateFormat(DATE_FORMAT_DAY_OF_WEEK, Locale.getDefault()).format(time)

```

File: app/src/main/res/values/dimens.xml [hunk 5]
```diff
@@ -1,12 +1,20 @@
...
     <dimen name="dp_80">80dp</dimen>
     <dimen name="dp_1">1dp</dimen>
+    <dimen name="dp_24">24dp</dimen>
+    <dimen name="dp_100">100dp</dimen>
+    <dimen name="dp_32">32dp</dimen>
+    <dimen name="dp_56">56dp</dimen>
+    <dimen name="scale_2">2</dimen>
+    <dimen name="sp_28">28sp</dimen>
+    <dimen name="dp_20">20dp</dimen>
+    <dimen name="dp_180">180dp</dimen>
```
Comment (thonc-2135) [inline app/src/main/res/values/dimens.xml]: sắp xếp lại cho n theo thứ tự nhé, để loạn xạ v, có thể tác riêng dp vs sp ra cũng được

File: app/src/main/res/layout/dialog_checkin.xml (hunk 1 above)
Comment (toandv98) [reply app/src/main/res/layout/dialog_checkin.xml]: Quên mất ko sửa lại

File: app/src/main/java/com/edu/sun/oereminder/ui/timetracking/TimeRecordAdapter.kt (hunk 2 above)
Comment (toandv98) [reply app/src/main/java/com/edu/sun/oereminder/ui/timetracking/TimeRecordAdapter.kt:30]: Không được ô ơi, itemView với timeRecord khác nhau

File: app/src/main/java/com/edu/sun/oereminder/utils/DateTimeExtension.kt (hunk 4 above)
Comment (toandv98) [reply app/src/main/java/com/edu/sun/oereminder/utils/DateTimeExtension.kt:13]: Trường hợp ngày ko được
"which can lead to unchecked nullability issues"

File: app/src/main/java/com/edu/sun/oereminder/ui/timetracking/TimeTrackingContract.kt [hunk 6]
```diff
@@ -1,14 +1,23 @@
...
+        fun showOrHideFab(isShow: Boolean)
+
+        fun updateRecyclerView(timeRecords: List<TimeRecord>)
+
+        fun navigateToCheckIn()
     }
 
     interface Presenter : BasePresenter<View> {
 
+        fun load()
```
Comment (huytq-2422) [inline app/src/main/java/com/edu/sun/oereminder/ui/timetracking/TimeTrackingContract.kt:19]: load gì đây nhỉ

File: app/src/main/java/com/edu/sun/oereminder/ui/timetracking/TimeTrackingContract.kt [hunk 7]
```diff
@@ -1,14 +1,23 @@
...
+
+        fun onDateSelected(from: Long, to: Long)
```
Comment (huytq-2422) [inline app/src/main/java/com/edu/sun/oereminder/ui/timetracking/TimeTrackingContract.kt:21]:
```suggestion
        fun selectDate(from: Long, to: Long)
```

File: app/src/main/java/com/edu/sun/oereminder/ui/timetracking/TimeTrackingFragment.kt [hunk 8]
```diff
@@ -1,13 +1,23 @@
...
+import com.google.android.material.datepicker.CalendarConstraints
+import com.google.android.material.datepicker.MaterialDatePicker
+import kotlinx.android.synthetic.main.fragment_time_tracking.*
+import kotlinx.android.synthetic.main.toolbar_app.*
+import java.util.*
 
 class TimeTrackingFragment : BaseFragment<View, TimeTrackingPresenter>(), View {
 
+    private var picker: MaterialDatePicker<Pair<Long, Long>>? = null
+    private val timeTrackingAdapter = TimeRecordAdapter()
```
Comment (huytq-2422) [inline app/src/main/java/com/edu/sun/oereminder/ui/timetracking/TimeTrackingFragment.kt:19]:
```suggestion
    private val timeTrackingAdapter by lazy {
        TimeRecordAdapter()
    }
```

File: app/src/main/java/com/edu/sun/oereminder/ui/timetracking/TimeTrackingFragment.kt [hunk 9]
```diff
@@ -18,9 +28,57 @@ class TimeTrackingFragment : BaseFragment<View, TimeTrackingPresenter>(), View {
...
+            activity?.let { picker?.show(it.supportFragmentManager, picker.toString()) }
+        }
+
+        picker?.addOnPositiveButtonClickListener {
+            btnDatePicker.text = picker?.headerText
+            it.first?.let { from ->
+                it.second?.let { to ->
+                    presenter.onDateSelected(from, to)
+                }
+            }
```
Comment (huytq-2422) [inline app/src/main/java/com/edu/sun/oereminder/ui/timetracking/TimeTrackingFragment.kt:65]: Nếu phải check nhiều thì không nên lồng nhau như thế 

```suggestion
            if(it.first != null && it.second != null) {
            }
```

File: app/src/main/java/com/edu/sun/oereminder/ui/timetracking/TimeTrackingContract.kt (hunk 6 above)
Comment (toandv98) [reply app/src/main/java/com/edu/sun/oereminder/ui/timetracking/TimeTrackingContract.kt:19]: load setupView, mà t ko biết đặt là gì cho hợp lý

File: app/src/main/java/com/edu/sun/oereminder/ui/timetracking/TimeRecordAdapter.kt [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+                    textTimeIn.text = timeInCalendar.toTime()
+                    textTimeOut.text = timeOutCalendar.toTime()
+                    textDate.text = dateCalendar.toDate()
+                    textStatus.text = if (status.isBlank()) {
+                        context.getString(R.string.status_pending)
+                    } else status
+                    textPartOfDay.text = partOfDay
+                }
+            }
+        }
```
Comment (huytq-2422) [inline app/src/main/java/com/edu/sun/oereminder/ui/timetracking/TimeRecordAdapter.kt:40]:
```suggestion
        fun bind(timeRecord: TimeRecord) = with(itemView) {
            with(timeRecord) {
                textTimeIn.text = timeInCalendar.toTime()
                textTimeOut.text = timeOutCalendar.toTime()
                textDate.text = dateCalendar.toDate()
                textStatus.text = if (status.isBlank()) {
                      context.getString(R.string.status_pending)
                } else status
                textPartOfDay.text = partOfDay
            }
        }
```

**Pull request merged**